cargo-quickinstall/src/lib.rs

```diff
@@ -129,141 +143,112 @@ pub fn do_dry_run_curl(crate_details: &CrateDetails) -> String {
         target = crate_details.target
     );
     if curl_head(&crate_download_url).is_ok() {
-        let cargo_bin_dir = home::cargo_home().unwrap().join("bin");
-        let cargo_bin_dir_str = cargo_bin_dir.to_str().unwrap();
-        format!(
-            "{curl_cmd:?} | {untar_cmd:?}",
-            curl_cmd = prepare_curl_bytes_cmd(&crate_download_url),
-            untar_cmd = prepare_untar_cmd(cargo_bin_dir_str)
-        )
+        let cargo_bin_dir = get_cargo_bin_dir()?;
+
+        Ok(format!(
+            "{curl_cmd} | {untar_cmd}",
+            curl_cmd = prepare_curl_bytes_cmd(&crate_download_url).formattable(),
+            untar_cmd = prepare_untar_cmd(&cargo_bin_dir).formattable(),
+        ))
     } else {
         let cargo_install_cmd = prepare_cargo_install_cmd(crate_details);
-        format!("{:?}", cargo_install_cmd)
+        Ok(format!("{}", cargo_install_cmd.formattable()))
     }
 }
 
-fn untar(tarball: Vec<u8>) -> Result<String, InstallError> {
-    use std::io::Write;
+fn untar(mut curl: ChildWithCmd) -> Result<String, InstallError> {
+    let bin_dir = get_cargo_bin_dir()?;
 
-    if tarball.is_empty() {
-        panic!("We fetched a tarball, but it was empty. Please report this as a bug.");
-    }
-    let cargo_home = home::cargo_home().unwrap();
-    let bin_dir = cargo_home.join("bin");
-    let mut tar_command = std::process::Command::new("tar");
-    tar_command
-        .arg("-xzvvf")
-        .arg("-")
-        .arg("-C")
-        .arg(bin_dir)
-        .stdin(std::process::Stdio::piped())
-        .stdout(std::process::Stdio::piped())
-        .stderr(std::process::Stdio::piped());
-    let mut process = tar_command.spawn()?;
-
-    process.stdin.take().unwrap().write_all(&tarball).unwrap();
-
-    let output = process.wait_with_output()?;
-
-    let stdout = String::from_utf8(output.stdout).unwrap();
-    let stderr = String::from_utf8(output.stderr).unwrap();
+    let res = prepare_untar_cmd(&bin_dir)
+        .stdin(curl.stdout().take().unwrap())
+        .output_checked_status();
 
-    if !output.status.success() {
-        return Err(CommandFailed {
-            command: "tar -xzvvf - -C ~/.cargo/bin".to_string(),
-            stdout,
-            stderr,
-        }
-        .into());
-    }
+    // Propagate this error before Propagating error tar since
+    // if tar fails, it's likely due to curl failed.
+    curl.wait_with_output_checked_status()?;
```

Comment:
it's hard to see how you're going to handle the `err.is_curl_404()` case under this scheme.

Reply:
In `untar`, it will propagate error from `curl` first, which includes the 404 error, before propagating error from `untar`.